sam.c

```diff
@@ -4955,6 +4955,164 @@ char *sam_open_mode_opts(const char *fn,
+            else if (strcmp(name, "MREVERSE") == 0) return BAM_FMREVERSE;
+            break;
+
+        case 'P':
+            if (strcmp(name, "PAIRED") == 0) return BAM_FPAIRED;
+            else if (strcmp(name, "PROPER_PAIR") == 0) return BAM_FPROPER_PAIR;
+            break;
+
+        case 'Q':
+            if (strcmp(name, "QCFAIL") == 0) return BAM_FQCFAIL;
+            break;
+
+        case 'R':
+            if (strcmp(name, "REVERSE") == 0) return BAM_FREVERSE;
+            else if (strcmp(name, "READ1") == 0) return BAM_FREAD1;
+            else if (strcmp(name, "READ2") == 0) return BAM_FREAD2;
+            break;
+
+        case 'S':
+            if (strcmp(name, "SECONDARY") == 0) return BAM_FSECONDARY;
+            else if (strcmp(name, "SUPPLEMENTARY") == 0) return BAM_FSUPPLEMENTARY;
+            break;
+
+        case 'U':
+            if (strcmp(name, "UNMAP") == 0) return BAM_FUNMAP;
+            break;
+        }
+    }
+
+    return short_ok? flag : 0;
+}
+
+// Parses one colon-terminated mode prefix, if present.
+static char parse_optional_mode_prefix(const char *s, char **endptr)
+{
+    char prefix[8];
+    int i = 0;
+
+    for (; isalpha_c(*s) && i < sizeof prefix; s++)
+        prefix[i++] = tolower_c(*s);
+
+    if (*s != ':') return '\0';
+
+    *endptr = (char *) &s[1];
+
+    if (i < sizeof prefix) {
+        prefix[i] = '\0';
+        switch (prefix[0]) {
+        case 'a':
+            if (strcmp(prefix, "all") == 0) return '&';
+            else if (strcmp(prefix, "any") == 0) return '|';
+            else if (strcmp(prefix, "anynot") == 0) return '~';
+            break;
+
+        case 'e':
+            if (strcmp(prefix, "exact") == 0) return '=';
+            break;
+
+        case 'n':
+            if (strcmp(prefix, "none") == 0) return '^';
+            else if (strcmp(prefix, "notall") == 0) return '~';
+            break;
+        }
+    }
+
+    return '\0';
+}
+
+int sam_parse_flag_filter(const char *s, char mode, sam_flag_filter_t *filt)
+{
+    while (*s) {
+        if (isalnum_c(*s) || *s == '_') {
+            char *end;
+            char newmode = parse_optional_mode_prefix(s, &end);
+            if (newmode) mode = newmode, s = end;
+            unsigned long flag = sam_parse_flag1(s, &end);
+            if (!flag) return -1;
+            s = end;
+
+            switch (mode) {
+            case '-': case '&':
+                filt->has_all = 1;
+                filt->all |= flag;
+                break;
+            case '!': case '^':
+                filt->has_none = 1;
+                filt->none |= flag;
+                break;
+            case '+': case '|':
+                filt->has_any = 1;
+                filt->any |= flag;
+                break;
+            case '~':
+                filt->has_omit_any = 1;
+                filt->omit_any |= flag;
+                break;
+            case '=':
+                filt->has_exact = 1;
+                filt->exact |= flag;
```

Comment:
See above comment - this is where I think it can simply set the `all` and `none` variables instead.